subsys/bluetooth/host/conn.c

```diff
@@ -1006,17 +942,33 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
 		return;
 	}
 
-	hdr = (void *)buf->data;
-	len = sys_le16_to_cpu(hdr->len);
+	if (conn->rx->len < sizeof(uint16_t)) {
+		/* Still not enough data recieved to retrieve the L2CAP header
+		 * length field.
+		 */
+		return;
+	}
 
-	if (sizeof(*hdr) + len != buf->len) {
-		BT_ERR("ACL len mismatch (%u != %u)", len, buf->len);
-		net_buf_unref(buf);
+	const struct bt_l2cap_hdr *hdr = (struct bt_l2cap_hdr *)conn->rx->data;
```

Comment:
Not a big fan of variable declaration in the middle of the function scope, also note that bt_l2cap_hdr contains the cid as well not just the len, so Id would suggest to just skip parsing the header and make bt_l2cap_recv do that using the whole buffer and if the header was not received (lenght <= 1) we should just drop.

Reply:
Thanks for taking the time to do another review round 👍 

> note that bt_l2cap_hdr contains the cid as well not just the len, so Id would suggest to just skip parsing the header and make bt_l2cap_recv do that using the whole buffer and if the header was not received (lenght <= 1) we should just drop.

I don't understand what you're asking me to do here. I'm not touching the `cid`, only using the `struct bt_l2cap_hdr` for retrieving the length field and for its size when calculating the total length of the ACL data. We need to parse the length field to know when to propagate the buffer to L2CAP, right? You're not saying you want to split this responsibility of reassembling across here and L2CAP? If you don't like the fact that the `struct bt_l2cap_hdr` type is used, it is possible to do something like

```c
const uint16_t l2cap_len = sys_get_le16(buf->data);
const uint16_t acl_total_len = l2cap_len + sizeof(struct bt_l2cap_hdr);
```
Regarding 
> Not a big fan of variable declaration in the middle of the function scope

Note that we cannot retrieve this length until we know that `buf->len >= 2`, which is not the case at the beginning of the function. Using C89-style would then force us to drop the `const` qualifier to declare a variable that's only used 60+ lines below. If we do strictly follow C89-style, then that's fine, I can declare the variable at the top of the function.